st2api/st2api/controllers/exp/inquiries.py

```diff
@@ -0,0 +1,295 @@
+        # existing_inquiry = inquiry_execution.__json__()
+        existing_inquiry = self._transform_inquiry(inquiry_execution)
+
+        if not requester_user:
+            requester_user = UserDB(cfg.CONF.system_user.user)
+
+        # Determine permission of this user to respond to this Inquiry
+        if not self._can_respond(existing_inquiry, requester_user):
+            abort(
+                http_client.FORBIDDEN,
+                'Insufficient permission to respond based on Inquiry parameters.'
+            )
+            return
+
+        # Validate the body of the response against the schema parameter for this inquiry
+        schema = existing_inquiry.get('schema')
+        LOG.debug("Validating inquiry response: %s against schema: %s" %
+                  (response_data.response, schema))
+        try:
+            jsonschema.validate(response_data.response, schema)
+        except jsonschema.exceptions.ValidationError:
+            abort(http_client.BAD_REQUEST, 'Response did not pass schema validation.')
+            return
+
+        # Update inquiry for completion
+        liveaction_db = self._mark_inquiry_complete(
+            inquiry_execution.liveaction.get('id'),
+            {"response": response_data.response}
+        )
+
+        # We only want to request a workflow resume if this has a parent
+        if liveaction_db.context.get("parent"):
+
+            # Request that root workflow resumes
+            root_liveaction = action_service.get_root_liveaction(liveaction_db)
+            action_service.request_resume(
+                root_liveaction,
+                requester_user
+            )
+
+        return {
+            "id": inquiry_id,
+            "response": response_data.response
+        }
+
+    def _transform_inquiry(self, raw_inquiry):
+        """Transform ActionExecutionAPI model into something specific to Inquiries
+
+            We're borrowing the ActionExecution data model for the time being, so we
+            need to pick and choose fields from this to form a new franken-model for
+            Inquiries so we don't return a bunch of extra crap. The idea is to
+            provide data in response to requests that make it look like Inquiry
+            is its own data model.
+        """
+
+        # The "put" and "get" commands will have an actual ActionExecutionAPI object, so
+        # if that's what's being passed in, let's convert it to a dict first.
+        # ("get_all" provides a dict natively)
+        if isinstance(raw_inquiry, ActionExecutionAPI):
+            raw_inquiry = raw_inquiry.__json__()
+
+        desired_fields = ["tag", "ttl", "users", "roles", "schema"]
+
+        inquiry = {
+            "id": raw_inquiry["id"]
+        }
+        for field in desired_fields:
+            inquiry[field] = raw_inquiry["result"][field]
+
+        return inquiry
+
+    def _mark_inquiry_complete(self, inquiry_id, result):
+        """Mark Inquiry as completed
```

Comment:
It looks like there is a small possible race in the "PUT" method (if multiple client send payload to that API endpoint in a short time frame) because we don't do any locking there, but that's probably fine, for now.

Reply:
Ack that. Will leave it for now but will keep this in mind if I hear about any issues. I'd assume the two client payloads would have to arrive in pretty close succession for this to happen